nixos/modules/services/security/bitwarden-directory-connector.nix

```diff
@@ -0,0 +1,343 @@
+{
+  pkgs,
+  config,
+  lib,
+  ...
+}:
+with lib; let
+  cfg = config.services.bitwarden_directory_connector;
+
+  ldap_data = builtins.toJSON cfg.ldap;
+  sync_data = builtins.toJSON cfg.sync;
+
+  # coping directly from nix https://github.com/NixOS/nixpkgs/blob/da4024d0ead5d7820f6bd15147d3fe2a0c0cec73/nixos/modules/config/nix.nix#L62C1-L76C49
+  semanticConfType = with types; let
+    confAtom =
+      nullOr
+      (oneOf [
+        bool
+        int
+        float
+        str
+        path
+        package
+      ])
+      // {
+        description = "Nix config atom (null, bool, int, float, str, path or package)";
+      };
+  in
+    attrsOf (either confAtom (listOf confAtom));
+
+  json_string = string: builtins.replaceStrings ["\""] ["\\\""] string;
+in {
+  imports = [];
+
+  options.services.bitwarden_directory_connector = {
+    enable = mkEnableOption "Bitwarden Directory Connector";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.bitwarden-directory-connector;
+      defaultText = literalExpression "pkgs.bitwarden-directory-connector";
+      description = lib.mdDoc "Reference to the Bitwarden Directory Connector package";
+      example = literalExpression "pkgs.bitwarden-directory-connector-example";
+    };
+
+    binary_name = mkOption {
+      type = types.str;
+      description = lib.mdDoc "The main binary for the connector.";
+      default = "bitwarden-directory-connector";
+    };
+
+    domain = mkOption {
+      type = types.str;
+      description = lib.mdDoc "The domain the Bitwarden/Vaultwarden is accessable on.";
+      example = "https://vaultwarden.example.com";
+    };
+
+    user = mkOption {
+      type = types.str;
+      description = lib.mdDoc "User to run the program.";
+      default = "bwdc";
+    };
+
+    directory = mkOption {
+      type = types.str;
+      description = lib.mdDoc "Folder to store the config file.";
+      default = "/etc/bitwarden/bwdc";
+    };
+
+    pw_env = mkOption {
+      type = types.str;
+      description = lib.mdDoc "The ENV var that the ldap password is stored.";
+      default = "LDAP_PW";
+    };
+    interval = mkOption {
+      type = types.str;
+      default = "*:0,15,30,45";
+      description = lib.mdDoc "When to run the connector, OnCalendar syntax.";
+    };
+
+    ldap = mkOption {
+      description = lib.mdDoc "Options to configurate LDAP.";
+      type = types.submodule {
+        freeformType = semanticConfType;
```

Comment:
The whole `semanticConfType` thing is specific to the nix settings usage. Sorry, I probably should have explained more clearly.

`freeformType` is essentially setting a type for the whole submodule, so you can allow things to be set outside of the declared options so long as the whole submodule fits that type in the end.

Here you'd want it to be `freeformType = pkgs.formats.json {};` (or possibly `freeformType = types.attrsOf (pkgs.formats.json {});` if it doesn't like the first version. `freeformType` might need it to be an attrs at the top level. I don't remember.)

Don't forget to remove the definition above, too.

Reply:
Tried both of those out and neither worked so I did a search on other modules.  
Found  ``freeformType = with types; attrsOf anything;`` and that works as I would expect it to.

Reply:
It's not a restrictive enough type. What went wrong with my suggestions?

Reply:
testing it with this config

```
      ldap = {
        password = "not_password";
        not_password = {
          active = false;
          thing = 1234;
          password = "password";
        };
      };
```


``freeformType = pkgs.formats.json {};``  
``error: A definition for option `services.bitwarden_directory_connector.ldap._module.freeformType' is not of type `null or optionType'. Definition values:``


``freeformType = types.attrsOf (pkgs.formats.json {});``  
`` error: attribute 'check' missing``

Reply:
Oops, it should be `freeformType = (pkgs.formats.json {}).type;`.

Reply:
That is done now.  
A heck of a lot cleaner all round